apax/data/input_pipeline.py

```diff
@@ -358,88 +360,82 @@ def next_power_of_two(x):
+
+    Parameters:
+    - value (int): The integer to round up.
+    - multiple (int): The multiple to round up to.
+
+    Returns:
+    - int: The rounded-up value.
+    """
+    return int(np.ceil(value / multiple) * multiple)
+
+
 class BatchProcessor:
-    def __init__(self, cutoff, forces=True, stress=False) -> None:
+    def __init__(
+        self, cutoff, atom_padding: int, nl_padding: int, forces=True, stress=False
+    ) -> None:
         self.cutoff = cutoff
+        self.atom_padding = atom_padding
+        self.nl_padding = nl_padding
+
         self.forces = forces
         self.stress = stress
 
     def __call__(self, samples: list[dict]):
+        n_samples = len(samples)
+        max_atoms = np.max([inp[0]["n_atoms"] for inp in samples])
+        max_atoms = round_up_to_multiple(max_atoms, self.atom_padding)
+
         inputs = {
-            "numbers": [],
-            "n_atoms": [],
-            "positions": [],
-            "box": [],
-            "idx": [],
-            "offsets": [],
+            "numbers": np.zeros((n_samples, max_atoms), dtype=np.int16),
+            "n_atoms": np.zeros(n_samples, dtype=np.int16),
+            "positions": np.zeros((n_samples, max_atoms, 3), dtype=np.float64),
+            "box": np.zeros((n_samples, 3, 3), dtype=np.float32),
         }
 
         labels = {
-            "energy": [],
+            "energy": np.zeros(n_samples, dtype=np.float64),
         }
 
         if self.forces:
-            labels["forces"] = []
-        if self.stress:
-            labels["stress"] = []
-
-        for sample in samples:
-            inp, lab = sample
-
-            inputs["numbers"].append(inp["numbers"])
-            inputs["n_atoms"].append(inp["n_atoms"])
-            inputs["positions"].append(inp["positions"])
-            inputs["box"].append(inp["box"])
-            idx, offsets = compute_nl(inp["positions"], inp["box"], self.cutoff)
-            inputs["idx"].append(idx)
-            inputs["offsets"].append(offsets)
-
-            labels["energy"].append(lab["energy"])
-            if self.forces:
-                labels["forces"].append(lab["forces"])
-            if self.stress:
-                labels["stress"].append(lab["stress"])
+            labels["forces"] = np.zeros((n_samples, max_atoms, 3), dtype=np.float64)
 
-        max_atoms = np.max(inputs["n_atoms"])
-        max_nbrs = np.max([idx.shape[1] for idx in inputs["idx"]])
+        idxs = []
+        offsets = []
+        for i, (inp, lab) in enumerate(samples):
+            inputs["numbers"][i, : inp["n_atoms"]] = inp["numbers"]
+            inputs["n_atoms"][i] = inp["n_atoms"]
+            inputs["positions"][i, : inp["n_atoms"]] = inp["positions"]
+            inputs["box"][i] = inp["box"]
 
-        max_atoms = next_power_of_two(max_atoms)
-        max_nbrs = next_power_of_two(max_nbrs)
+            idx, offset = compute_nl(inp["positions"], inp["box"], self.cutoff)
```

Comment:
You do not pass the `n_atoms` here so `compute_nl` computes the full neighbourlist, even if some positions are `zero` from padding? Wouldn't it be saver to not zero-pad but use NaN and filter? Is that slower? Can it be an issue if an atom is exactly at `(0, 0, 0)` which can happen quite frequently?

Reply:
`inp["positions"]` are the unpadded positions. Padded positions are masked during model calculations not by their 0,0,0 value, but by the atomic number 0 in the `numbers` array.